File: Standalone/NowUI.Bridge/Recorder.cs
// W2 - the managed half of the transport. Docs/Standalone/M3-Spec.md sections 5.1, 5.2, 5.4, 5.5.
//
// Four things live here and nothing else:
//
//   1. the managed buffers, and the growth protocol that resizes them without ever losing a frame (section 5.1);
//   2. the preamble - the intern-table pass that must run BEFORE decoding, once, so the decode itself allocates
//      nothing that outlives the call and is therefore idempotent under exactLayout (section 5.6);
//   3. the validator of section 5.5, which refuses a malformed buffer with a slot offset;
//   4. string access for the replay: an intern handle or a volatile-table index, resolved the same way.
//
// It does NOT call JavaScript. The one [JSImport] lives in BridgeInterop.cs, and `Record` takes it as a delegate,
// so Standalone/NowUI.Bridge.Tests can drive the whole protocol - growth, preamble and validation - against a
// hand-built frame with no browser, no wasm and no GPU.

using System;
using System.Collections.Generic;
using System.Text;

namespace NowUI.Bridge
{
    /// <summary>
    /// A malformed command buffer. Section 5.5: the decoder validates the whole buffer before it opens a single
    /// scope, and a failure names the slot offset. NowUI is never touched when this is thrown.
    /// </summary>
    public sealed class NowBridgeProtocolException : Exception
    {
        public NowBridgeProtocolException(int slot, string message)
            : base("NowUI bridge: malformed command buffer at slot " + slot + ". " + message)
        {
            Slot = slot;
        }

        /// <summary>The slot offset the failure was found at. -1 when the failure is not slot-addressed.</summary>
        public int Slot { get; }
    }

    /// <summary>
    /// What one boundary crossing looks like from the managed side. Section 5.1's <c>record</c>, as a delegate so
    /// the protocol can be tested without a browser.
    /// </summary>
    public delegate int RecordFunc(
        Span<int> ops,
        Span<byte> opsText,
        Span<int> results,
        Span<byte> resultsText,
        int resultSlots,
        int frame);

    /// <summary>The managed buffers, the growth protocol, the preamble and the validator.</summary>
    public sealed class BridgeRecorder
    {
        // Sized so a first frame of a small application fits without a single growth, and so the two-slot
        // NEED_MORE write always has somewhere to land. Section 5.1: "ops is always >= 2 slots, so this always
        // fits" - that is a promise this side keeps, not one the JavaScript side can check.
        private const int InitialOpSlots = 1024;
        private const int InitialTextBytes = 4096;

        /// <summary>Builds a recorder over a fresh result table.</summary>
        public BridgeRecorder()
            : this(null)
        {
        }

        /// <summary>Builds a recorder over <paramref name="results"/>, so a host can hold the table too.</summary>
        public BridgeRecorder(BridgeResults results)
        {
            m_ResultsTable = results ?? new BridgeResults();
        }

        private int[] m_Ops = new int[InitialOpSlots];
        private byte[] m_OpsText = new byte[InitialTextBytes];

        /// <summary>
        /// W5's table (section 6). Owned by <see cref="BridgeResults"/> rather than by this class, because the two
        /// directions have different growth rules: the op stream grows through the NEED_MORE handshake because
        /// JavaScript writes it, and the result table grows by a plain array copy because the managed side writes
        /// it between crossings.
        /// </summary>
        private readonly BridgeResults m_ResultsTable;

        /// <summary>Section 5.4: the wasm side keeps a List&lt;string&gt; at the same indices JavaScript's Map uses.</summary>
        private readonly List<string> m_Strings = new List<string>(256);

        // This frame's volatile table, as (byteOffset, byteLength) pairs into m_OpsText.
        private int[] m_VolatileOffsets = new int[64];
        private int[] m_VolatileLengths = new int[64];
        private int m_VolatileCount;

        private int m_Used;
        private int m_OpStart;
        private int m_OpEnd;
        private int m_FrameFlags;

        /// <summary>Total slots the last <see cref="Record"/> brought back, header included.</summary>
        public int used => m_Used;

        /// <summary>First slot of the op stream.</summary>
        public int opStart => m_OpStart;

        /// <summary>One past the last op slot.</summary>
        public int opEnd => m_OpEnd;

        /// <summary>Header slot 2. See <see cref="Abi.FlagFaulted"/> and friends.</summary>
        public int frameFlags => m_FrameFlags;

        /// <summary>Section 4.4: the author's draw function threw, and the buffer is a balanced prefix.</summary>
        public bool faulted => (m_FrameFlags & Abi.FlagFaulted) != 0;

        /// <summary>Section 5.7: whether this frame asks for the two-pass measured replay.</summary>
        public bool exactLayout => (m_FrameFlags & Abi.FlagExactLayout) != 0;

        /// <summary>How many strings have been interned this session. Diagnostics and tests.</summary>
        public int internedCount => m_Strings.Count;

        /// <summary>W5's table (section 6). The replay writes it; <see cref="Record"/> carries it back.</summary>
        public BridgeResults results => m_ResultsTable;

        /// <summary>The whole frame, header included. Valid only after <see cref="Preamble"/> returned.</summary>
        public ReadOnlySpan<int> frame => new ReadOnlySpan<int>(m_Ops, 0, m_Used);

        /// <summary>The op stream on its own.</summary>
        public ReadOnlySpan<int> ops => new ReadOnlySpan<int>(m_Ops, m_OpStart, m_OpEnd - m_OpStart);

        /// <summary>Raw slot access for the decoder, which walks absolute slot indices.</summary>
        public int Slot(int index) => m_Ops[index];

        /// <summary>Reinterprets a slot as an f32 (section 5.3's <c>f32</c> kind).</summary>
        public float SlotF32(int index) => BitConverter.Int32BitsToSingle(m_Ops[index]);

        // -------------------------------------------------------------------------------------------- record

        /// <summary>
        /// One boundary crossing, plus at most one growth retry. Section 5.1.
        /// </summary>
        /// <remarks>
        /// The retry does not re-run the author's draw function: the JavaScript side keeps the recorded bytes and
        /// its <c>pending</c> flag, and answers the second call with two set()s. So no handler fires twice and no
        /// state is read twice - which is the property that makes growth free rather than a dropped frame.
        /// <para>Two retries would mean the JavaScript side asked for a size and then did not fit in it, which is a
        /// bridge bug rather than a growth, and is reported as one.</para>
        /// </remarks>
        public int Record(RecordFunc record, int frameNumber)
        {
            if (record == null) throw new ArgumentNullException(nameof(record));

            for (int attempt = 0; attempt < 2; ++attempt)
            {
                // The result table's spans are re-read from the table on every attempt, because a growth between
                // frames replaces the arrays. `sealedSlots` - not the buffer length - is what JavaScript slices:
                // it is the header plus this frame's records, so a table that grew once to hold a long string
                // does not cost a full-buffer copy on every frame after it.
                int returned = record(
                    m_Ops.AsSpan(),
                    m_OpsText.AsSpan(),
                    m_ResultsTable.slots.AsSpan(),
                    m_ResultsTable.text.AsSpan(),
                    m_ResultsTable.sealedSlots,
                    frameNumber);

                if (returned != Abi.NeedMore)
                {
                    if (returned < 0)
                        throw new NowBridgeProtocolException(-1,
                            "record() returned " + returned + ", and the only negative return is NEED_MORE (" +
                            Abi.NeedMore + ").");

                    if (returned > m_Ops.Length)
                        throw new NowBridgeProtocolException(-1,
                            "record() reported " + returned + " slots but the buffer it was given holds only " +
                            m_Ops.Length + ".");

                    m_Used = returned;
                    return returned;
                }

                if (attempt == 1)
                    throw new NowBridgeProtocolException(-1,
                        "record() asked to grow twice for one frame. The JavaScript side reported a size and then " +
                        "did not fit in it; this is a bridge bug, not a growth.");

                // Section 5.1: the two sizes the managed side must grow to are in the first two slots.
                int needSlots = m_Ops[0];
                int needText = m_Ops[1];

                if (needSlots <= 0 || needSlots > Abi.MaxSlots)
                    throw new NowBridgeProtocolException(0,
                        "NEED_MORE asked for " + needSlots + " op slots, which is outside 1.." + Abi.MaxSlots + ".");

                if (needText < 0)
                    throw new NowBridgeProtocolException(1,
                        "NEED_MORE asked for " + needText + " text bytes.");

                Grow(ref m_Ops, needSlots);
                Grow(ref m_OpsText, needText);
            }

            throw new InvalidOperationException("unreachable");
        }

        private static void Grow(ref int[] buffer, int need)
        {
            if (need <= buffer.Length) return;
            int size = buffer.Length;
            while (size < need) size *= 2;
            buffer = new int[size];
        }

        private static void Grow(ref byte[] buffer, int need)
        {
            if (need <= buffer.Length) return;
            int size = buffer.Length;
            while (size < need) size *= 2;
            buffer = new byte[size];
        }

        // ------------------------------------------------------------------------------------------ preamble

        /// <summary>
        /// Validates the whole buffer (section 5.5) and interns this frame's new strings, in that order, before a
        /// single scope is opened. Throws <see cref="NowBridgeProtocolException"/> naming a slot offset on any
        /// failure, with NowUI untouched.
        /// </summary>
        /// <remarks>
        /// The interning happens HERE rather than in the decoder, and that is what makes the decoder idempotent
        /// (section 5.6): under exactLayout the buffer is decoded twice per frame, and a decoder that interned as
        /// it went would append the same strings twice on the second pass.
        /// </remarks>
        public void Preamble(int usedSlots)
        {
            m_Used = usedSlots;
            Validate(usedSlots);
            InternNewStrings();
        }

        /// <summary>The validator on its own, for the tests that assert what each rule refuses.</summary>
        public void Validate(int usedSlots)
        {
            // ---- rule 1: magic and surface hash -----------------------------------------------------------
            if (usedSlots < Abi.HeaderSlots)
                throw new NowBridgeProtocolException(0,
                    "the frame is " + usedSlots + " slots long and the fixed header alone is " + Abi.HeaderSlots + ".");

            int magic = m_Ops[Abi.HdrMagic];
            if (magic != Abi.Magic)
                throw new NowBridgeProtocolException(Abi.HdrMagic,
                    "magic is 0x" + magic.ToString("X8") + ", expected 0x" + Abi.Magic.ToString("X8") + ".");

            int surfaceHash = m_Ops[Abi.HdrSurfaceHash];
            if (surfaceHash != Abi.SurfaceHash)
                throw new NowBridgeProtocolException(Abi.HdrSurfaceHash,
                    "this wasm build's surface hash is 0x" + Abi.SurfaceHash.ToString("X8") +
                    " and nowui.js sent 0x" + surfaceHash.ToString("X8") +
                    ". The JavaScript bundle and the wasm module were built from different surface manifests; " +
                    "rebuild both, or serve the nowui.js that matches this module.");

            // ---- rule 2: the header's own numbers are in bounds and mutually consistent --------------------
            m_FrameFlags = m_Ops[Abi.HdrFrameFlags];
            int internCount = m_Ops[Abi.HdrInternCount];
            int volatileCount = m_Ops[Abi.HdrVolatileCount];
            int textBytes = m_Ops[Abi.HdrTextBytes];
            int start = m_Ops[Abi.HdrOpStart];
            int end = m_Ops[Abi.HdrOpEnd];

            if (internCount < 0 || volatileCount < 0)
                throw new NowBridgeProtocolException(internCount < 0 ? Abi.HdrInternCount : Abi.HdrVolatileCount,
                    "a table count is negative (internCount " + internCount + ", volatileCount " + volatileCount + ").");

            if (textBytes < 0 || textBytes > m_OpsText.Length)
                throw new NowBridgeProtocolException(Abi.HdrTextBytes,
                    "textBytes is " + textBytes + " and the text buffer holds " + m_OpsText.Length + ".");

            long expectedStart = (long)Abi.HeaderSlots + (long)internCount * 3 + (long)volatileCount * 2;
            if (start != expectedStart)
                throw new NowBridgeProtocolException(Abi.HdrOpStart,
                    "opStart is " + start + " but the header plus " + internCount + " intern triples and " +
                    volatileCount + " volatile pairs ends at " + expectedStart + ".");

            if (end < start || end > usedSlots)
                throw new NowBridgeProtocolException(Abi.HdrOpEnd,
                    "opEnd is " + end + ", which is outside opStart (" + start + ") .. used (" + usedSlots + ").");

            // ---- rule 3: intern declarations are contiguous from the current count, and in the text buffer --
            int cursor = Abi.HeaderSlots;
            int expectedHandle = m_Strings.Count;

            for (int i = 0; i < internCount; ++i, cursor += 3)
            {
                int handle = m_Ops[cursor];
                int offset = m_Ops[cursor + 1];
                int length = m_Ops[cursor + 2];

                if (handle != expectedHandle + i)
                    throw new NowBridgeProtocolException(cursor,
                        "intern declaration " + i + " has handle " + handle + "; declarations must be contiguous " +
                        "from this session's current intern count (" + expectedHandle + "), so it should be " +
                        (expectedHandle + i) + ".");

                if (offset < 0 || length < 0 || (long)offset + length > textBytes)
                    throw new NowBridgeProtocolException(cursor + 1,
                        "intern declaration " + i + " spans bytes " + offset + ".." + (offset + length) +
                        " of a text section that is " + textBytes + " bytes long.");
            }

            for (int i = 0; i < volatileCount; ++i, cursor += 2)
            {
                int offset = m_Ops[cursor];
                int length = m_Ops[cursor + 1];

                if (offset < 0 || length < 0 || (long)offset + length > textBytes)
                    throw new NowBridgeProtocolException(cursor,
                        "volatile string " + i + " spans bytes " + offset + ".." + (offset + length) +
                        " of a text section that is " + textBytes + " bytes long.");
            }

            // ---- rule 4: walking argSlots from opStart lands exactly on opEnd -----------------------------
            // ---- rule 5: scope opens and closes balance and nest -------------------------------------------
            //
            // One walk does both, because a desynchronised walk would make the balance count meaningless anyway.
            int depth = 0;
            int slot = start;

            while (slot < end)
            {
                int header = m_Ops[slot];
                int opcode = header & 0xFFFF;
                int argSlots = (header >> 16) & 0xFFFF;

                if (opcode == Abi.OpInvalid)
                    throw new NowBridgeProtocolException(slot,
                        "opcode 0 (OP_INVALID) is never emitted; a zeroed buffer fails here.");

                long next = (long)slot + 1 + argSlots;
                if (next > end)
                    throw new NowBridgeProtocolException(slot,
                        "the op at this slot declares " + argSlots + " argument slots, which runs to " + next +
                        " and past opEnd (" + end + "). The buffer is truncated or the op stream is desynchronised.");

                OpSpec spec = Abi.Find(opcode);

                // A fixed op must be exactly its declared width; a variable one (strlist, opts - section 5.3)
                // must be at least its minimum, because its real width lives in the count or bitmask slot the
                // decoder has not read yet. Rule 4's walk stays exact either way: the width is taken from the
                // op header, not from this table.
                if (spec != null && (spec.Variable ? argSlots < spec.Slots : argSlots != spec.Slots))
                    throw new NowBridgeProtocolException(slot,
                        "op " + spec.Name + " (opcode " + opcode + ") declares " + argSlots +
                        " argument slots and this build expects " +
                        (spec.Variable ? "at least " : "") + spec.Slots + ".");

                if (opcode == Abi.OpScopeClose)
                {
                    if (--depth < 0)
                        throw new NowBridgeProtocolException(slot,
                            "OP_SCOPE_CLOSE with no scope open. Scope opens and closes must balance and nest.");
                }
                else if (spec != null && spec.OpensScope)
                {
                    ++depth;
                }

                slot = (int)next;
            }

            if (slot != end)
                throw new NowBridgeProtocolException(slot,
                    "walking the op stream landed on slot " + slot + " and opEnd is " + end + ".");

            if (depth != 0)
                throw new NowBridgeProtocolException(end,
                    depth + " scope(s) were opened and never closed. The recorder closes every scope in a finally " +
                    "block, so an unbalanced stream is a bridge bug rather than an author error.");

            // Cached for the decode, now that they are known good.
            m_OpStart = start;
            m_OpEnd = end;
            LoadVolatileTable(internCount, volatileCount);
        }

        private void LoadVolatileTable(int internCount, int volatileCount)
        {
            if (m_VolatileOffsets.Length < volatileCount)
            {
                int size = m_VolatileOffsets.Length;
                while (size < volatileCount) size *= 2;
                m_VolatileOffsets = new int[size];
                m_VolatileLengths = new int[size];
            }

            int cursor = Abi.HeaderSlots + internCount * 3;
            for (int i = 0; i < volatileCount; ++i, cursor += 2)
            {
                m_VolatileOffsets[i] = m_Ops[cursor];
                m_VolatileLengths[i] = m_Ops[cursor + 1];
            }

            m_VolatileCount = volatileCount;
        }

        private void InternNewStrings()
        {
            int internCount = m_Ops[Abi.HdrInternCount];
            int cursor = Abi.HeaderSlots;

            for (int i = 0; i < internCount; ++i, cursor += 3)
            {
                int offset = m_Ops[cursor + 1];
                int length = m_Ops[cursor + 2];
                m_Strings.Add(Encoding.UTF8.GetString(m_OpsText, offset, length));
            }
        }

        // -------------------------------------------------------------------------------------------- strings

        /// <summary>
        /// Section 5.3's <c>str</c> kind: a non-negative slot is an intern handle, a negative one is the bitwise
        /// complement of an index into this frame's volatile table. An interned string costs nothing to read; a
        /// volatile one costs one allocation per distinct string per frame, which is what keeps a text field from
        /// leaking an intern-table entry per keystroke.
        /// </summary>
        public string Text(int slot)
        {
            if (slot >= 0)
            {
                if (slot >= m_Strings.Count)
                    throw new NowBridgeProtocolException(-1,
                        "intern handle " + slot + " was used and only " + m_Strings.Count +
                        " strings have been declared this session.");

                return m_Strings[slot];
            }

            int index = ~slot;
            if (index >= m_VolatileCount)
                throw new NowBridgeProtocolException(-1,
                    "volatile string index " + index + " was used and this frame declared " + m_VolatileCount + ".");

            return Encoding.UTF8.GetString(m_OpsText, m_VolatileOffsets[index], m_VolatileLengths[index]);
        }

        /// <summary>An interned string by handle, for diagnostics that render a path segment.</summary>
        public string Interned(int handle)
        {
            return handle >= 0 && handle < m_Strings.Count ? m_Strings[handle] : null;
        }

        /// <summary>
        /// Drops every interned string. Only a hot reload calls this (W7): the JavaScript side's Map and this list
        /// are two halves of one table, and resetting one without the other is how handles come to mean different
        /// strings on the two sides.
        /// </summary>
        public void ResetStrings()
        {
            m_Strings.Clear();
        }
    }
}

File: Assets/NowUI/WebBundle~/nowui/recorder.js
// W2 - the recorder. Docs/Standalone/M3-Spec.md sections 5.1 to 5.4.
//
// JavaScript owns its buffers. This file is the whole of that ownership: a growable ArrayBuffer with an Int32Array,
// a Float32Array and a Uint8Array over it, the two string tables, the op stream, and the header that describes
// them. The recorder writes at full speed into its own memory with no MemoryView proxy anywhere in the path; the
// only crossing is the pair of set() calls in bridge.js, once per frame.
//
// Why the op stream is recorded into a second buffer and assembled at flush. Section 5.2 puts the intern and
// volatile tables BETWEEN the fixed header and the op stream, and neither table's length is known until the frame
// is over - a string is interned when it is first used, which can be the last op. Recording into a separate
// buffer and copying it into place at flush costs one memcpy of the op stream (about 7 KB for a realistic frame,
// against the ~350 KB of vertex data the WebGL2 backend already moves in the same frame) and keeps the wire format
// exactly as specified. The alternative - a second set() at a target offset - depends on MemoryView.set accepting
// an offset argument, which W1 measures but which nothing else in the design needs; this does not bet on it.

import {
    MAGIC, HDR_SLOTS, HDR_MAGIC, HDR_SURFACE_HASH, HDR_FRAME_FLAGS, HDR_INTERN_COUNT, HDR_VOLATILE_COUNT,
    HDR_TEXT_BYTES, HDR_OP_START, HDR_OP_END,
    FLAG_FAULTED, FLAG_EXACT_LAYOUT, FLAG_HAS_NEW_STRINGS,
    OP_SCOPE_CLOSE, MAX_SLOTS, DEFAULT_MAX_STRINGS, SURFACE_HASH,
} from './abi.js';

import { PathTrie, NowUIAuthorError } from './trie.js';
import { Results } from './results.js';

const encoder = new TextEncoder();

export class Recorder {
    constructor(options = {}) {
        this.maxSlots = options.maxSlots || MAX_SLOTS;
        this.maxStrings = options.maxStrings || DEFAULT_MAX_STRINGS;
        this.exactLayout = options.exactLayout !== false;

        // The op stream, recorded here and copied into `out` at flush.
        this._allocOps(1024);
        this.opsUsed = 0;

        // The assembled frame: header + tables + op stream. What crosses the boundary.
        this.out = new Int32Array(1024);
        this.used = 0;

        // UTF-8 bytes for this frame's new interns and this frame's volatile strings, in that order.
        this.u8 = new Uint8Array(4096);
        this.textUsed = 0;

        // Section 5.4. handle -> permanent for the session; the wasm side keeps a List<string> at the same indices.
        this.interned = new Map();
        this.internCount = 0;
        this.newInterns = [];       // flat triples: handle, byteOffset, byteLength
        this.volatiles = [];        // flat pairs:   byteOffset, byteLength

        this.frame = 0;
        this.faulted = false;
        this.faultError = null;
        this.pending = false;       // section 5.1: a NEED_MORE retry re-uses the bytes already recorded

        this.openScopes = [];       // labels, for the cap message and for closeAll
        this.overflowed = false;

        this.trie = new PathTrie(key => this.intern(key), {
            maxStrings: this.maxStrings,
            onReport: options.onReport,
        });

        // W5's table (section 6). It lives on the recorder because the two directions are one conversation: a
        // value control resolves its answer out of `results` and emits the resolved value into `ops` in the same
        // call (section 6.4), and separating the two would mean the surface holds two objects that must never
        // disagree about which frame it is.
        this.results = new Results({
            debug: options.debug === true,
            onReport: options.onResultReport,
        });
    }

    // ------------------------------------------------------------------------------------------ frame bracket

    /// Starts a new recording. Called only when `pending` is false, so a growth retry never re-runs the author's
    /// draw function (section 5.1) and no handler fires twice.
    reset(frame) {
        this.frame = frame;
        this.opsUsed = 0;
        this.textUsed = 0;
        this.newInterns.length = 0;
        this.volatiles.length = 0;
        this.openScopes.length = 0;
        this.faulted = false;
        this.faultError = null;
        this.overflowed = false;
        this.trie.beginFrame(frame);
    }

    /// Section 4.4. The author's draw function threw. The frame is marked faulted and whatever was recorded before
    /// the throw is kept: it is a balanced prefix, because closeAll() runs on every path out.
    fault(error) {
        this.faulted = true;
        this.faultError = error;
    }

    /// Section 4.1: "the buffer is balanced on every path". Closes every scope the draw function left open - which
    /// only happens when it threw, since every scope helper closes in a finally.
    closeAll() {
        while (this.openScopes.length > 0) this.endScope();
        this.trie.endFrame();
    }

    // ------------------------------------------------------------------------------------------ emitting

    /// Reserves one op header slot plus `slots` argument slots, and writes the header. Callers then write exactly
    /// `slots` argument slots with i32/f32. `argSlots` is carried in the high 16 bits so a decoder meeting an
    /// unknown opcode can skip it rather than desynchronise (section 5.2).
    op(record, slots = record.slots) {
        this._ensureOps(this.opsUsed + 1 + slots);
        this.ops[this.opsUsed++] = (record.opcode & 0xffff) | ((slots & 0xffff) << 16);
    }

    i32(value) {
        this.ops[this.opsUsed++] = value | 0;
    }

    f32(value) {
        this.opsF32[this.opsUsed++] = value;
    }

    /// Opens a scope: emits its op, then remembers it so the matching OP_SCOPE_CLOSE is emitted even if the body
    /// throws. `node` is the trie node the scope pushed; its label is what the cap message prints.
    beginScope(record, node, writeArgs) {
        this.op(record);
        writeArgs();
        this.openScopes.push(node.label);
    }

    endScope() {
        this.openScopes.pop();

        // Once the slot cap has been hit the buffer is being abandoned, and asking to grow it again would throw a
        // SECOND time - out of the finally block that closeAll() runs in, replacing the cap message (which names
        // the open scopes) with an identical one that names none, because they have just been popped. Measured,
        // not reasoned: that is exactly what the first run of the cap test produced.
        if (!this.overflowed) {
            this._ensureOps(this.opsUsed + 1);
            this.ops[this.opsUsed++] = OP_SCOPE_CLOSE & 0xffff;   // argSlots 0
        }

        this.trie.pop();
    }

    // ------------------------------------------------------------------------------------------ strings

    /// Section 5.4, interned. A string is interned on first use and its handle is permanent for the session.
    /// Handles introduced this frame are declared in the header's intern table, contiguously from the wasm side's
    /// current count, so the decoder adds them in one pass before it decodes anything.
    intern(text) {
        const existing = this.interned.get(text);
        if (existing !== undefined) return existing;

        if (this.internCount >= this.maxStrings) {
            // Section 5.4's ceiling. Reached through a KEY, this throws: a UI minting unbounded distinct keys is
            // deriving identity from data, and stopping is correct. Reached through a plain string, the caller
            // (`str`) catches this and falls back to the volatile path with one warning.
            throw new NowUIAuthorError(
                'NowUI: the string intern table is full (' + this.maxStrings + ' entries).\n' +
                'Keys are interned permanently, so this means keys are being derived from data that changes -\n' +
                'a timestamp, a counter, an interpolated value. Give the control a stable key and pass the\n' +
                'changing text as an option: ui.button(\'retry\', { label: `Retry (${n})` }).');
        }

        const handle = this.internCount++;
        const { offset, length } = this._writeText(text);
        this.newInterns.push(handle, offset, length);
        this.interned.set(text, handle);
        return handle;
    }

    /// Section 5.3's `str` argument kind. A string the recorder has seen before interns; anything else this frame
    /// is volatile - written into opsText for this frame only, and referenced as ~index so the sign discriminates
    /// the two tables with no extra slot.
    str(text) {
        if (text === null || text === undefined) text = '';
        else if (typeof text !== 'string') text = String(text);

        const existing = this.interned.get(text);
        if (existing !== undefined) return existing;

        if (this.internCount < this.maxStrings && this._shouldIntern(text)) {
            return this.intern(text);
        }

        const { offset, length } = this._writeText(text);
        this.volatiles.push(offset, length);
        return ~(this.volatiles.length / 2 - 1);
    }

    /// Section 5.4's volatile path, taken unconditionally. For a control's VALUE - a text field's contents, a
    /// formatted number - the caller knows the string is per-frame data and says so, rather than leaving it to
    /// the heuristic below.
    ///
    /// W5 is where this had to exist, and the reason is worth writing down because the heuristic looks right
    /// until a value control is drawn through it.
    ///
    /// Section 5.4 says the volatile path is "the difference between a text field costing one UTF-8 encode per
    /// keystroke and permanently leaking one intern-table entry per keystroke", and that a `str` position
    /// "interns only when the same string instance or value has been seen before". Those two are inconsistent for
    /// a text field, and the inconsistency is only visible once one exists. Type "Ada": frame N emits "A", which
    /// has not been seen, so it is volatile - and then frame N+1 emits "A" AGAIN, because the author echoes the
    /// value back through section 6.4's rule, and "seen before" is now true. Every prefix of every string the user
    /// types is interned one frame after it is typed.
    ///
    /// Measured in the browser before this method existed: typing "Ada Lovelace" into wwwroot/nowui/app.stub.js's
    /// w5 field took the session's intern count from 12 to 34 - twenty-two permanent handles for twelve
    /// keystrokes - and clearing and retyping took it to 43. At section 5.4's ceiling of 65 536 that is a few
    /// minutes of typing before a key throws.
    volatileStr(text) {
        if (text === null || text === undefined) text = '';
        else if (typeof text !== 'string') text = String(text);

        const { offset, length } = this._writeText(text);
        this.volatiles.push(offset, length);
        return ~(this.volatiles.length / 2 - 1);
    }

    /// The recorder decides intern-vs-volatile by argument position (the op table's `str` kind) and by whether the
    /// value has been seen before - never by guessing at the content. `seen` is a per-session set of values that
    /// have appeared in a `str` position at least once; the second sighting interns.
    ///
    /// It is the right rule for a LABEL, which is what `str` positions mostly are: a label built by interpolation
    /// (`'Retry (' + n + ')'`) stays volatile while n changes and interns once it settles. It is the wrong rule
    /// for a VALUE, which is why `volatileStr` exists above and why the surface calls it for one.
    _shouldIntern(text) {
        if (this._seen === undefined) this._seen = new Set();
        if (this._seen.has(text)) { this._seen.delete(text); return true; }
        if (this._seen.size > 4096) this._seen.clear();
        this._seen.add(text);
        return false;
    }

    _writeText(text) {
        const offset = this.textUsed;

        // Worst case is 4 bytes per UTF-16 code unit; asking for that up front means encodeInto never has to run
        // twice, and the buffer stays allocated for the session.
        this._ensureText(offset + text.length * 4 + 4);

        const result = encoder.encodeInto(text, this.u8.subarray(offset));
        this.textUsed = offset + result.written;
        return { offset, length: result.written };
    }

    // ------------------------------------------------------------------------------------------ flush

    /// Assembles header + tables + op stream into `out`, and sets `used` / `textUsed` to the two sizes the managed
    /// buffers must be able to hold. Section 5.2's layout, in order.
    flush() {
        const internCount = this.newInterns.length / 3;
        const volatileCount = this.volatiles.length / 2;
        const opStart = HDR_SLOTS + internCount * 3 + volatileCount * 2;
        const opEnd = opStart + this.opsUsed;

        this._ensureOut(opEnd);
        const out = this.out;

        out[HDR_MAGIC] = MAGIC;
        out[HDR_SURFACE_HASH] = SURFACE_HASH;
        out[HDR_FRAME_FLAGS] =
            (this.faulted ? FLAG_FAULTED : 0) |
            (this.exactLayout ? FLAG_EXACT_LAYOUT : 0) |
            (internCount > 0 ? FLAG_HAS_NEW_STRINGS : 0);
        out[HDR_INTERN_COUNT] = internCount;
        out[HDR_VOLATILE_COUNT] = volatileCount;
        out[HDR_TEXT_BYTES] = this.textUsed;
        out[HDR_OP_START] = opStart;
        out[HDR_OP_END] = opEnd;

        let cursor = HDR_SLOTS;
        for (let i = 0; i < this.newInterns.length; i++) out[cursor++] = this.newInterns[i];
        for (let i = 0; i < this.volatiles.length; i++) out[cursor++] = this.volatiles[i];

        out.set(this.ops.subarray(0, this.opsUsed), opStart);

        this.used = opEnd;
        return this;
    }

    // ------------------------------------------------------------------------------------------ growth

    _allocOps(slots) {
        const buffer = new ArrayBuffer(slots * 4);
        this.ops = new Int32Array(buffer);
        this.opsF32 = new Float32Array(buffer);
    }

    _ensureOps(need) {
        if (need <= this.ops.length) return;

        if (need > this.maxSlots) {
            this.overflowed = true;
            const tail = this.openScopes.slice(-8);
            throw new NowUIAuthorError(
                'NowUI: the draw function emitted more than ' + this.maxSlots.toLocaleString('en-US') +
                ' command slots; this is almost always an unbounded loop or a recursive component. ' +
                'The last ' + tail.length + ' scope keys were: ' + (tail.length ? tail.join(', ') : '<none>') + '.');
        }

        let size = this.ops.length;
        while (size < need) size *= 2;
        if (size > this.maxSlots) size = this.maxSlots;

        const buffer = new ArrayBuffer(size * 4);
        const grown = new Int32Array(buffer);
        grown.set(this.ops.subarray(0, this.opsUsed));
        this.ops = grown;
        this.opsF32 = new Float32Array(buffer);
    }

    _ensureOut(need) {
        if (need <= this.out.length) return;
        let size = this.out.length;
        while (size < need) size *= 2;
        this.out = new Int32Array(size);
    }

    _ensureText(need) {
        if (need <= this.u8.length) return;
        let size = this.u8.length;
        while (size < need) size *= 2;
        const grown = new Uint8Array(size);
        grown.set(this.u8.subarray(0, this.textUsed));
        this.u8 = grown;
    }
}
